---
title: "Single-copy genes missing and present from Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

### data cleaning
Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(vroom)
library(GENESPACE)
setwd("~/Dropbox/rumex_pangenome_annotation")
```

#### raw files
Loading in wide format from query_pangenes() for eah possible reference and bed files.
```{r echo=TRUE, message=FALSE}
# excluded non representative array members and non-syntenic othogroups (see genespace script)
hap1_raw<-data.table::fread("pg_hap1_wide_synonly.txt")
hap2_raw<-data.table::fread("pg_hap2_wide_synonly.txt")
sal_raw<-data.table::fread("pg_sal_wide_synonly.txt")

hap1_bed<-vroom("hap1.bed", col_names = F)
## use hap1 as the total? since that's the ref?
## and hap1 ref vs sal as ref gives same number of Y loss genes anyway

## total is number of pgs? 
#47360 pgIDs - all have position equivs in hap1
# less in sal - some of those may not have the right position info
bla<-setdiff(sal_raw$salicifolius,hap1_raw$salicifolius) # 2450 genes in sal not in hap1.
## ok genes vs pgIDs
head(bla) # not in b
## use salicifolius gene name, and pgID as key - relative to hap1?
hap1_raw %>% select(salicifolius,hap1) %>% n_distinct() ## ok so we've got some duplicate entires but each pgID is distinct
pgIDkey<-hap1_raw %>% select(pgID,interpChr,start,end,hap1,salicifolius) # ?? useful?
#47360
hap1_raw %>%select(pgID) %>% n_distinct() #
#47360

hap1_raw %>% select(interpChr,start,end, interpOrd) %>% n_distinct()
#47285 have distinct start,end,chr combos - only discernable via interpOrd column
#47360 only reached once interpOrd column added

# hap1_total<-hap1_raw %>% filter(salicifolius !="") #19882 entries
# hap1_total_names <- hap1_total %>% select(hap1) %>% filter(hap1!="")
# hap1_totalb<- hap1_raw %>%filter(hap1!="") %>% filter(salicifolius!="")
# sal_total<-sal_raw %>%  filter(salicifolius !="") #22126
# #sal_total_names <- sal_total %>% select(hap1) %>%filter(hap1 !="")
# ##
# #bla<-as.data.frame(setdiff(sal_total$salicifolius,hap1_total$salicifolius)) #22116 shared
# ## 2450 genes no in hap1. should these be considered in the total? well no in my opinion since im defininf this by total gene number wrt to location in hap1 and i DONT KNOW the location in hap1 
# ??????
# k
## but for plotting don't i just want genes w a hap1 equiv?? 

```

### Gene loss on Y 
Identify genes that are shared by salicifolius and hap1, but not hap2.
```{r}
hap1_yloss<-hap1_raw %>% filter(hap1!="" & hap2=="" & salicifolius!="")  #%>%select(salicifolius) %>% n_distinct()#1337 distincT!!
  ## ok two duplicates present here but not in the sal ref - not a big deal imo 
#sal_yloss<-sal_raw %>% filter(hap1!="" & hap2=="" & salicifolius!="") #%>% select(salicifolius) %>% n_distinct()#1337 distinctgenes
#hap2_yloss<- hap2_raw%>% filter(hap1!="" & hap2=="" & salicifolius!="") #%>% select(salicifolius) %>% n_distinct()#956 distinct genes

# hap1sal<-union(hap1_yloss$salicifolius,sal_yloss$salicifolius) # hap1 and sal ref files have same list of y loss genes 
# #1337 shared
# hap1sal<-as.data.frame(hap1sal) ## post filtered set
# # hap1hap2<-union(hap1_yloss$salicifolius,hap2_yloss$salicifolius) # hap2 and hap1 have 381 distinct y loss genes
# #1337 shared
# hap1hap2<-as.data.frame(hap1hap2)
# hap2sal<-union(hap2_yloss$salicifolius,sal_yloss$salicifolius) #hap2 and sal ref have same list of y loss genes
# #1337 shared
# #hap2sal<-as.data.frame(hap2sal)
# ## to do: incorporate list of Y loss genes from hap1, which are more numerous than those present in the 
# # going with the shared set of 1337 genes i guess!

```
#### connect to positions
```{r}

write.csv(hap1_yloss, "hap1sal_nohap2.csv")
```
### Summary of gene list

#### summarising number of genes
```{r}
a<-hap1_yloss %>% group_by(interpChr) %>% summarise(n=n()) %>% 
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

b<-hap1_raw %>% group_by(interpChr) %>% summarise(n=n()) %>%
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")
# b<-hap1pg %>% filter(!is.na(hap1)) %>% group_by(interpChr) %>% summarise(n=n()) %>%
#   filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

# ## ok what about the counts from the og bed file
bedbedbed<-vroom("hap1.bed", col_names = F)
c<-bedbedbed %>% rename(interpChr=X1) %>% group_by(.,interpChr) %>% summarise(n=n())  %>%
   filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")
```
#### How many genes are missing from theX compared to other chr?
What proportion of genes (wrt hap1 bed file) have a syntenic ortholog shared with Salicifolius but NOT hap2?
```{r proportion}
chrpropn<-full_join(b,a,by="interpChr",suffix=c(".total",".notonhap2")) %>% 
  mutate(., propn=n.notonhap2/n.total) %>% mutate(.,percent = propn*100)
## 
```
#### PARs and old and new sex-linked regions?
```{r}
yloss_pars <- hap1_yloss %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_yloss_pars <-yloss_pars %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

total_pangenes<-hap1_raw %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_total_pangenes <-total_pangenes %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

# bedfix<-bedbedbed %>% rename(interpChr=X1)%>% rename(end=X3) %>%
#   mutate(.,region=if_else(interpChr =="X" &end<=71000000,"PAR1",
#                         if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
#                         if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
#                         if_else(interpChr=="X" & end>367000000,"PAR2",
#                         if_else(interpChr=="A1","A1",
#                         if_else(interpChr=="A2","A2",
#                         if_else(interpChr=="A4","A4",NA))))))))
# 
# c<-bedfix %>% group_by(region) %>% summarise(n=n())  %>%
#   filter(!is.na(region))
# ## keep only non dup genes
# b<-bedfix%>% filter(X4%in%hap1nodups$id) %>% group_by(region) %>% summarise(n=n())  %>%
#   filter(!is.na(region))
 chrpropnPAR<-full_join(summary_total_pangenes,summary_yloss_pars,by="region",suffix=c(".total",".notonhap2")) %>% 
   mutate(., propn=n.notonhap2/n.total) %>% mutate(.,percent = propn*100) %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 
## plot proprtions tho

ggplot(chrpropnPAR) + geom_col(aes(x=region,y=propn)) + 
  theme_classic() + 
  ylab("proportion annotated pangenes\nshared by haplotype 1 and salicifolius\n with no haplotype 2 ortholog")
ggsave("pangenepropn_yloss.png", width=6,height=4) 

yloss.vector<-as.vector(chrpropnPAR$n.notonhap2)
total.vector<-as.vector(chrpropnPAR$n.total)
proptesttest<-prop_test(yloss.vector,total.vector )
proptesttest$df
as.character(chrpropnPAR$region)
xtab<-select(chrpropnPAR,region,n.total,n.notonhap2) %>% column_to_rownames(var="region")
#rownames(xtab)<-xtab$region
dimnames(xtab) 
row_wise_prop_test(xtab)
xq<-chisq.test(xtab,p=rep(1/7,7))
xq$p.value
```
#### results so far
genes missing from hap2 
```{r}
library(knitr)
kable(chrpropnPAR)

zoestuff<-filter(sal_raw,hap1!="",hap2!="",salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
zoestuff2<-filter(hap1_raw,hap1!="",hap2!="",salicifolius!="") %>% select(hap1,hap2,salicifolius) %>% n_distinct()
zoestuff3<-filter(hap1_raw,hap1!="",hap2!="",salicifolius!="") %>% select(hap1,hap2,salicifolius) %>% n_distinct()

write.csv(zoestuff, "syntenic_orthos_all.csv",quote = F,row.names = F)

```

###redoing bc i messed up the denominator (again!!!)
```{r}
hasanortho<-hap1_raw %>% filter((hap1!=""|hap2!="")&salicifolius!="")

hap1_yloss_redo<-hasanortho %>% filter(hap1!="" & hap2=="" & salicifolius!="")  #number doesn't change (expected)

a_redo<-hap1_yloss_redo %>% group_by(interpChr) %>% summarise(n=n()) %>% 
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

b_redo<-hasanortho %>% group_by(interpChr) %>% summarise(n=n()) %>%
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

yloss_pars_redo <- hap1_yloss_redo %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_yloss_pars_redo <-yloss_pars_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

total_pangenes_redo<-hasanortho %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_total_pangenes_redo <-total_pangenes_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

chrpropnPAR_redo<-full_join(summary_total_pangenes_redo,summary_yloss_pars_redo,by="region",suffix=c(".total",".notonhap2")) %>% 
   mutate(., propn=n.notonhap2/n.total) %>% mutate(.,percent = propn*100) %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 
## plot proprtions tho

ggplot(chrpropnPAR_redo) + geom_col(aes(x=region,y=propn)) + 
  theme_classic() + 
  ylab("proportion annotated pangenes\nshared by haplotype 1 and salicifolius\n with no haplotype 2 ortholog")
ggsave("pangenepropn_yloss_redo.png", width=6,height=4) 
```
#### list for blasting genes against hap2 to confirm their loss??

```{r}
genelisth1<-hap1_yloss_redo %>% select(hap1)
genelistsal<-hap1_yloss %>% select(salicifolius)
#write.table(genelisth1,"genelisthap1.txt",quote=FALSE,sep="\t")
#write.table(genelistsal, "genelistsal.txt",quote=FALSE,sep="\t")


```


#### plotting
```{r}
#library(ggbio)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(tidyverse)
# bedanno<-import("hap1.bed")
# head(bedanno)
# 
# 
# (n(hap1raw$ofID))


```



