---
title: "Presence/absence variation among Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

# data cleaning
Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(vroom)
```

## raw files
Load in long-format pangenome databases for each reference
```{r echo=TRUE, message=FALSE}
hap1raw<-vroom("hap1_pangenes.txt", show_col_types = F)
hap2raw<-vroom("hap2_pangenes.txt",show_col_types=F)
salraw<-vroom("salicifolius_pangenes.txt", show_col_types=F)
```
What do these column names mean?

colname  |definition
---------|------
ofID | orthofinder-given unique gene ID
pgID | the unique position-by-orthogroup combination identifier. Each row in the pangenome-annotation is a unique value of this field
interpChr |  the chromosome (syntenic or actual) on the reference genome
interpOrd     | the gene-rank order position (interpolated or actual) on the reference genome
pgRepID     | the ofID for the representative gene for each pgID
genome    | genome of the focal gene (id column)
og   | orthogroup ID taken from the combBed.txt file
flag| specifies whether the gene is a syntenic ortholog (‘PASS’), non-syntenic ortholog, or array member.
id| bed-specified gene ID 
chr| chromosome
start| start pos
end | end pos
ord | gene rank order

### what do I need to do?
```{r}
hap1syn<- hap1raw %>% dplyr::filter(flag=="PASS")
hap2syn<-hap2raw %>% dplyr::filter(flag=="PASS")
salsyn<-salraw %>% dplyr::filter(flag=="PASS")
```
Determine which genes are present in haplotype 1 and present or absent in haplotype 2
Determine which genes are present in haplotype 2 and not presetn in absent in haplotype 1
Determine which genes are not shared by sal and the other two, or are only shared between sal and one or the other. 
### what will the data look like
Here I'm envisioning a table of shared positions and gene names (column separated by species. I can include a flag indication which don't have a syntenic ortholog. 
simplest candidate for Y gene loss: single copy syntenic hap1 and sal but not in hap2.
What is my ultimate question? 
Is there gene loss on Y? neo Y? 
### Gene loss on Y

#### remove dups and combine to complete list
```{r}
hap1nodups_a<-hap1syn %>% filter(genome == "hap1") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
hap1nodups_b<- hap2syn %>%filter(genome == "hap1") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
hap1nodups_c<-salsyn %>%filter(genome == "hap1") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
hap1nodups<-union(hap1nodups_a,hap1nodups_b) 
hap1nodups<-union(hap1nodups,hap1nodups_c) 
##
hap2nodups_a<-hap1syn %>% filter(genome=="hap2") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
hap2nodups_b<-hap2syn %>%filter(genome == "hap2") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
hap2nodups_c<-salsyn %>%filter(genome == "hap2") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
hap2nodups<-union(hap2nodups_a,hap2nodups_b) 
hap2nodups<-union(hap2nodups,hap2nodups_c)
##
salnodups_a<-hap1syn %>% filter(genome=="salicifolius") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
salnodups_b<-hap2syn %>% filter(genome=="salicifolius") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
salnodups_c<-salsyn %>% filter(genome=="salicifolius") %>% distinct(pgID,.keep_all=TRUE) %>% select(id)
salnodups<-union(salnodups_a,salnodups_b) 
salnodups<-union(salnodups,salnodups_c)
###
```


#### Syntenic genes
Obtain a list of genes shared by hap1, sal, but NOT hap2
```{r}
## remove dups and get into wide format
hap1pg<-hap1syn %>%select(pgID,genome,id) %>%
  filter((id%in%salnodups$id | id%in%hap2nodups$id | id%in%hap1nodups$id)) %>%
  pivot_wider(names_from=genome, values_from=id,values_fn=list) %>% 
  mutate_all( ~replace(., lengths(.)==0, NA))
hap2pg <-hap2syn %>%select(pgID,genome,id) %>%
  filter((id%in%salnodups$id | id%in%hap2nodups$id | id%in%hap1nodups$id)) %>%
  pivot_wider(names_from=genome, values_from=id,values_fn=list) %>% 
  mutate_all( ~replace(., lengths(.)==0, NA))
salpg<- salsyn %>%select(pgID,genome,id) %>%
  filter((id%in%salnodups$id | id%in%hap2nodups$id | id%in%hap1nodups$id)) %>%
  pivot_wider(names_from=genome, values_from=id,values_fn=list) %>% 
  mutate_all( ~replace(., lengths(.)==0, NA))
##
df1<-hap1pg %>% filter(!is.na(hap1) & is.na(hap2) & !is.na(salicifolius)) %>% select(hap1)
df2<-hap2pg %>% filter(!is.na(hap1) & is.na(hap2) & !is.na(salicifolius)) %>% select(hap1)
df3<-salpg %>% filter(!is.na(hap1) & is.na(hap2) & !is.na(salicifolius)) %>% select(hap1)
## combined list
all_nohap2<-intersect(df1,df2)
all_nohap2<-intersect(all_nohap2,df3)
## THESE ARE THE GENES W SAL AND HAP1 BUT NO HAP2 equivs
```
#### connect to positions
```{r}
all_nohap2$hap1<- as.character(all_nohap2$hap1)
all_nohap2
hap1syn
anno_pg<-inner_join(all_nohap2,hap1syn, by=c("hap1"="id"))
write.csv(anno_pg, "hap1sal_nohap2.csv")
```
#### plotting? or...
