---
title: "Single-copy genes missing and present from Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

### data cleaning
Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(vroom)
library(GENESPACE)
setwd("~/Dropbox/rumex_pangenome_annotation")
```

#### raw files
Loading in wide format from query_pangenes() for each possible reference and bed files.
```{r echo=TRUE, message=FALSE}
# excluded non representative array members and non-syntenic othogroups (see genespace script)
hap1_raw<-data.table::fread("pg_hap1_wide_synonly.txt")
hap2_raw<-data.table::fread("pg_hap2_wide_synonly.txt")
sal_raw<-data.table::fread("pg_sal_wide_synonly.txt")
hap1_nonsyn<-data.table::fread("pg_hap1_wide.txt")

hap1_bed<-vroom("hap1.bed", col_names = F)

```
### get rid of pgs w non-syn OG members
```{r}
nsorths<-hap1_nonsyn %>% filter_all(any_vars(grepl("\\*",.))) %>%  filter!grepl("\\|",)))#16071 ns orths 

onlysyn<-hap1_raw %>% filter(!(pgID%in%nsorths$pgID)) # 31289 genes remain?


onlysyn_sal<-sal_raw %>% filter(!(pgID%in%nsorths$pgID)) # 29906?

```

### Gene loss on Y 
Identify genes that are shared by salicifolius and hap1, but not hap2.
```{r}
hap1_yloss<-onlysyn %>% filter(hap1!="" & hap2=="" & salicifolius!="")
```
#### connect to positions
```{r}
write.csv(hap1_yloss, "hap1sal_nohap2_syn.csv")
```
### PARS and correct denominator
```{r}
hasanortho<-onlysyn %>% filter((hap1!=""|hap2!="")&salicifolius!="")

hap1_yloss_redo<-hasanortho %>% filter(hap1!="" & hap2=="" & salicifolius!="")  #number doesn't change (expected)

a_redo<-hap1_yloss_redo %>% group_by(interpChr) %>% summarise(n=n()) %>% 
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

b_redo<-hasanortho %>% group_by(interpChr) %>% summarise(n=n()) %>%
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

yloss_pars_redo <- hap1_yloss_redo %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_yloss_pars_redo <-yloss_pars_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

total_pangenes_redo<-hasanortho %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_total_pangenes_redo <-total_pangenes_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

chrpropnPAR_redo<-full_join(summary_total_pangenes_redo,summary_yloss_pars_redo,by="region",suffix=c(".total",".notonhap2")) %>% 
   mutate(., propn=n.notonhap2/n.total) %>% mutate(.,percent = propn*100) %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 
## plot proportions 

ggplot(chrpropnPAR_redo) + geom_col(aes(x=region,y=propn)) + 
  theme_classic() + 
  ylab("proportion annotated pangenes\nshared by haplotype 1 and salicifolius\n with no haplotype 2 ortholog")
ggsave("pangenepropn_yloss_redo.png", width=6,height=4) 
```

#### results so far
genes missing from hap2 
```{r}
library(knitr)
kable(chrpropnPAR_redo)
```

```{r}

#zoestuff<-filter(onlysyn,hap1!="",hap2!="",salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
## at least one
#zoestuff2<-filter(onlysyn_sal,(hap1!=""|hap2!="") &salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
#zoestuff2<-filter(onlysyn_sal,(hap1!=""|hap2!="") &salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
#write.csv(zoestuff2, "syntenic_orthos_sal.csv",quote = F,row.names = F)
#zoestuff3<-read.csv("pg_hap1_wide_onetoone.txt")
```
#### list for blasting genes against hap2 to confirm their loss??

```{r}
 genelisth1<-hap1_yloss_redo %>% select(hap1)
# genelistsal<-hap1_yloss %>% select(salicifolius)
# #write.table(genelisth1,"genelisthap1.txt",quote=FALSE,sep="\t")
# #write.table(genelistsal, "genelistsal.txt",quote=FALSE,sep="\t")
```


#### plotting
```{r}
```



