---
title: "Presence/absence variation among Rumex xyy male haplotype assemblies and Rumex salicifolius"
output: html_notebook
---

# data cleaning
The pangenome outputs from genespace are in long format, or a wider format which weird file separation. First attempt will use the long format and a tidy style approach.

Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r echo=TRUE}
library(tidyverse)
library(vroom)
```

## raw files
Load in long-format pangenome databases for each reference
```{r echo=TRUE}
hap1raw<-vroom("hap1_pangenomeDB.txt", show_col_types = F)
hap2raw<-vroom("hap2_pangenomeDB.txt",show_col_types=F)
salraw<-vroom("salicifolius_pangenomeDB.txt", show_col_types=F)
colnames(hap1raw)
```
What do these column names mean?

colname  |definition
---------|------
pgChr    |pangenome chromosome (relative to reference)
pgOrd    |gene rank order
pgID     |unique id for each entry
og       |orthogroup (as defined by orthofinder)
ofID     |orthofinder ID
isRep    |??
genome   |genome that gene is found in
isDirectSyn| is syntenic?
isSynOfOnly| is syntenic ortholog only?
isNSOrtho|is nonsyntenic ortholog
id| geneID in gff

hap1syn

## record scratch