---
title: "Single-copy genes missing and present from Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

### data cleaning
Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(vroom)
library(GENESPACE)
setwd("~/Dropbox/rumex_pangenome_annotation")
```
#### pre-cleaning using grep

#### raw files
Loading in wide format from query_pangenes() for each possible reference and bed files.
```{r echo=TRUE, message=FALSE}
# excluded non representative array members and non-syntenic othogroups (see genespace script)
hap1_raw<-data.table::fread("pg_hap1_wide_synonly.txt")
hap2_raw<-data.table::fread("pg_hap2_wide_synonly.txt")
sal_raw<-data.table::fread("pg_sal_wide_synonly.txt")
hap1_nonsyn<-data.table::fread("pg_hap1_wide.txt")

hap1_bed<-vroom("hap1.bed", col_names = F)

```
### get rid of pgs w non-syn OG members
```{r}
nsorths<-hap1_nonsyn %>% filter_all(any_vars(grepl("\\*",.))) 

onlysyn<-hap1_raw %>% filter(!(pgID%in%nsorths$pgID)) # 31289 genes remain?


onlysyn_sal<-sal_raw %>% filter(!(pgID%in%nsorths$pgID)) # 29906?

```

### Gene loss on Y 
Identify genes that are shared by salicifolius and hap1, but not hap2.
```{r}
#hap1_yloss<-onlysyn %>% filter(hap1!="" & hap2=="" & salicifolius!="")

#write.csv(hap1_yloss, "hap1sal_nohap2_syn.csv")
```
### PARS and correct denominator
```{r}
hasanortho<-onlysyn %>% filter((hap1!=""|hap2!="")&salicifolius!="")

hap1_yloss_redo<-hasanortho %>% filter(hap1!="" & hap2=="" & salicifolius!="")  #number doesn't change (expected)

a_redo<-hap1_yloss_redo %>% group_by(interpChr) %>% summarise(n=n()) %>% 
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

b_redo<-hasanortho %>% group_by(interpChr) %>% summarise(n=n()) %>%
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

yloss_pars_redo <- hap1_yloss_redo %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_yloss_pars_redo <-yloss_pars_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

total_pangenes_redo<-hasanortho %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_total_pangenes_redo <-total_pangenes_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

chrpropnPAR_redo<-full_join(summary_total_pangenes_redo,summary_yloss_pars_redo,by="region",suffix=c(".total",".notonhap2")) %>% 
   mutate(., propn=n.notonhap2/n.total) %>% mutate(.,percent = propn*100) %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 
## plot proportions 

ggplot(chrpropnPAR_redo) + geom_col(aes(x=region,y=propn)) + 
  theme_classic() + 
  ylab("proportion annotated pangenes\nshared by haplotype 1 and salicifolius\n with no haplotype 2 ortholog")
ggsave("pangenepropn_yloss_redo.png", width=6,height=4) 
```

#### results so far
genes missing from hap2 
```{r}
library(knitr)
kable(chrpropnPAR_redo)
```

```{r}

#zoestuff<-filter(onlysyn,hap1!="",hap2!="",salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
## at least one
#zoestuff2<-filter(onlysyn_sal,(hap1!=""|hap2!="") &salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
#zoestuff2<-filter(onlysyn_sal,(hap1!=""|hap2!="") &salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
#write.csv(zoestuff2, "syntenic_orthos_sal.csv",quote = F,row.names = F)
#zoestuff3<-read.csv("pg_hap1_wide_onetoone.txt")
```
#### list for blasting genes against hap2 to confirm their loss

```{r}
genelisth1<-hap1_yloss_redo %>% select(hap1)
#genelistsal<-hap1_yloss_redo %>% select(salicifolius)
write.table(genelisth1,"genelisthap1.txt",quote=FALSE,sep="\t")
# #write.table(genelistsal, "genelistsal.txt",quote=FALSE,sep="\t")
```
##### BLAST
Run in ohta because stuff is installed there. blastexploration.sh and blast_hap2loss.sh
```{bash eval=FALSE, warning=FALSE,error=FALSE}
#!/bin/bash
## parse fasta headers
awk -F"-" '/^>/{print $1; next}1' NChap1_final.all.maker.transcripts.fasta > NChap1.out.fasta
## get list of genes to blast in fasta format
awk '{print $2}' genelisthap1.txt > genelisthap1.out.txt 
echo "Filter out candidate loss genes from nchap1 transcript fasta"
genelist="genelisthap1.out.txt"
fasta="NChap1.out.fasta"
while read i; do
    samtools faidx $fasta $i  >> hap1_sequences.fa
done < $genelist

wait
## make blast db from nc hap2 transcripts
echo "starting blastn"
makeblastdb -in NChap2_final.all.maker.transcripts.fasta -dbtype nucl \
  -input_type fasta -out hap2transcripts

# the following are one-liners I used to parse and explore blast output

## blast results - blasting hap1 missing genes against hap2 transcript fasta ##
## first attempt:
blastn -db hap2transcripts -query hap1_sequences.fa   -out blastout.txt
blastn -db hap2transcripts -query hap1_sequences.fa  -outfmt 6 -out blastout_tab.txt
## in parsing the results I realized that there were multiple instances of blast hits for the same sequence pairs. Not very useful for me, because at least one strong blast his is enough for me to discount a gene a properly missing.
## setting max-hsps to 1 (default is 0, whihc means no limit)
blastn -db hap2transcripts -query hap1_sequences.fa -max_hsps 1  -outfmt 6 -out blast_max1.tsv
blastn -db hap2transcripts -query hap1_sequences.fa -max_hsps 1 -out blast_max1.xml.txt
## only the xml format reveals "no hits found" genes
grep -B 5 "No hits found" blast_max1.xml.txt |grep "Query" > nohits.max1.txt
## counting
less nohits.max1.txt |wc -l
#output: 400
awk '{print $2}' nohits.max1.txt | grep -v "|" >nohits.max1.genes.txt
# 397 - 3 of those were dups
##----------------------------------------------------------------------
## unique hap1 genes in blast output
awk '{print $1}' blast_max1.tsv | sort| uniq |wc -l
# 449, 122 entries are duplicate

## total genes in putative missing list 
#849

## numbers add up - out of the 849 genes, 400 have no hits, 449 have at least one hit, 122 of those have multiple hits in hap2 (on different genes)
##----------------------------------------------------------------------
awk '{ if ($3 < 90) { print } }' blast_max1.tsv > lowpercenthits_max1.txt
# 151 genes have low % ID hits
awk '{ if ($11 < 0.01) { print } }' blast_max1.tsv > goodEhits_max1.txt
# 663 (all hits) have good "e" lol
awk '{ if ($3 < 97) { print } }' blast_max1.tsv > reallylowpercenthits_max1.txt
# 240 genes have low ID % hits when we make it more stringent
```
#### futher analysis of blast output in R
problem: 
```{r, warning=FALSE,message=FALSE}
blastout<-read_tsv("blast_max1.tsv",
                   col_names = c("qseqid","sseqid","pident",
                                "length","mismatch","gapopen",
                                "qstart","qend","sstart","send","evalue","bitscore"))
blast<-blastout %>% select(qseqid,sseqid,pident,evalue)
blast_uniq<- blastout  %>% add_count(qseqid) %>% filter(n==1)
blast_dup<- blastout %>% add_count(qseqid) %>% filter(n>1)

blast_pos<-full_join(blastout,hap2_raw,by=c("qseqid"="hap1")) %>% select(qseqid,start,end,sseqid,interpChr,chr,pident) %>% add_count(qseqid) 
blast_uniq_pos<-blast_pos %>%filter(n==1)
blast_dup_pos<-blast_pos %>%filter(n>1) 

#note from stephen
#The catch is when there are paralogs in the genome, and how to sort those out- no significant match on Y1 or Y2 is one easy extra test, but then there might be significant matches on Y1 or Y2 that are weaker than on autosomes- I think in those cases we would most simply treat as losses as well
blast_post %>% 
```
e.g. NChap1_final_00000186 has 3 hits to hap2 - one with 100% id, one with 98, one with 83. all on X

#### plotting
```{r}
```



