---
title: "Single-copy genes missing and present from Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

### data cleaning
Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(vroom)
library(GENESPACE)
setwd("~/Dropbox/rumex_pangenome_annotation")
```
#### pre-cleaning using grep

#### raw files
Loading in wide format from query_pangenes() for each possible reference and bed files.
```{r echo=TRUE, message=FALSE}
# excluded non representative array members and non-syntenic othogroups (see genespace script)
hap1_raw<-data.table::fread("pg_hap1_wide_synonly.txt")
hap2_raw<-data.table::fread("pg_hap2_wide_synonly.txt")
sal_raw<-data.table::fread("pg_sal_wide_synonly.txt")
hap1_nonsyn<-data.table::fread("pg_hap1_wide.txt")

hap1_bed<-vroom("hap1.bed", col_names = F)

```
### get rid of pgs w non-syn OG members
```{r}
nsorths<-hap1_nonsyn %>% filter_all(any_vars(grepl("\\*",.))) 

onlysyn<-hap1_raw %>% filter(!(pgID%in%nsorths$pgID)) # 31289 genes remain?


onlysyn_sal<-sal_raw %>% filter(!(pgID%in%nsorths$pgID)) # 29906?

```

### Gene loss on Y 
Identify genes that are shared by salicifolius and hap1, but not hap2.
```{r}
#hap1_yloss<-onlysyn %>% filter(hap1!="" & hap2=="" & salicifolius!="")

#write.csv(hap1_yloss, "hap1sal_nohap2_syn.csv")
```
### PARS and correct denominator
```{r}
hasanortho<-onlysyn %>% filter((hap1!=""|hap2!="")&salicifolius!="")

hap1_yloss_redo<-hasanortho %>% filter(hap1!="" & hap2=="" & salicifolius!="")  #number doesn't change (expected)

a_redo<-hap1_yloss_redo %>% group_by(interpChr) %>% summarise(n=n()) %>% 
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

b_redo<-hasanortho %>% group_by(interpChr) %>% summarise(n=n()) %>%
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

yloss_pars_redo <- hap1_yloss_redo %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_yloss_pars_redo <-yloss_pars_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

total_pangenes_redo<-hasanortho %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_total_pangenes_redo <-total_pangenes_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

chrpropnPAR_redo<-full_join(summary_total_pangenes_redo,summary_yloss_pars_redo,by="region",suffix=c(".total",".notonhap2")) %>% 
   mutate(., propn=n.notonhap2/n.total) %>% mutate(.,percent = propn*100) %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 
## plot proportions 

ggplot(chrpropnPAR_redo) + geom_col(aes(x=region,y=propn)) + 
  theme_classic() + 
  ylab("proportion annotated pangenes\nshared by haplotype 1 and salicifolius\n with no haplotype 2 ortholog")
ggsave("pangenepropn_yloss_redo.png", width=6,height=4) 
```

#### results so far
genes missing from hap2 
```{r}
library(knitr)
kable(chrpropnPAR_redo)
```

```{r}

#zoestuff<-filter(onlysyn,hap1!="",hap2!="",salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
## at least one
#zoestuff2<-filter(onlysyn_sal,(hap1!=""|hap2!="") &salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
#zoestuff2<-filter(onlysyn_sal,(hap1!=""|hap2!="") &salicifolius!="") %>% select(hap1,hap2,salicifolius) #%>% n_distinct()
#write.csv(zoestuff2, "syntenic_orthos_sal.csv",quote = F,row.names = F)
#zoestuff3<-read.csv("pg_hap1_wide_onetoone.txt")
```
#### list for blasting genes against hap2 to confirm their loss

```{r}
genelisth1<-hap1_yloss_redo %>% select(hap1)
#genelistsal<-hap1_yloss_redo %>% select(salicifolius)
write.table(genelisth1,"genelisthap1.txt",quote=FALSE,sep="\t")
# #write.table(genelistsal, "genelistsal.txt",quote=FALSE,sep="\t")
```
##### BLAST
Run in ohta because stuff is installed there
```{bash eval=TRUE, error=FALSE, warning=FALSE}
# ./blast_genome.sh
cat blast_genome.sh
```
#### futher analysis of blast output in R

```{r, warning=FALSE,message=FALSE}
blastout<-read_tsv("blast_h2genome.tsv",
                    col_names = c("qseqid", "sseqid", "pident", "length", 
                                  "mismatch", "gapopen", "qstart", "qend", 
                                  "qlen", "sstart", "send", "slen", "evalue", 
                                  "bitscore", "score"))

blast_pos<-full_join(blastout,hap2_raw,by=c("qseqid"="hap1"))  %>% add_count(qseqid) 
### which genes, out of any, are no hits in hap2 genome
hits<-blastout %>% select(qseqid) %>% distinct()
nohit<- anti_join(genelisth1,hits, by = c("hap1"="qseqid")) # 117 genes in genelist do not have a hit
# not a surprise, its a big genome! and lots might be garbage!
okhit<- blastout %>% filter(pident > 90) #19537/25326 are 90% id
### for each gene with at least one hit - is the top hit on Y1 or Y2?
# group by gene and chromosome match, mutate ranking pid with highest value having rank 1
tophit<- blastout %>% group_by(qseqid, sseqid) %>%
  mutate(the_rank  = rank(-pident, ties.method = "random")) %>%
  filter(the_rank == 1) %>% select(-the_rank) %>% mutate(alignedpropn = length/qlen) 
dim(tophit) #1174 genes based on top hit per chr only
tophit_eval<- blastout %>% group_by(qseqid, sseqid) %>%
  mutate(the_rank  = rank(evalue, ties.method = "random")) %>%
  filter(the_rank == 1) %>% select(-the_rank) # pick the ones with the lowest e-value
setdiff(tophit$qseqid,tophit_eval$qseqid) # same! whether we ranked based on percent id or e val
tophitY<-filter(tophit,grepl("Y.",sseqid)) # get all the Y1 and Y2 hits

dim(tophitY) # 418 genes have a top hit on Y1 or Y2
dim(filter(tophitY, sseqid == "Y1")) # 198 on Y1
dim(filter(tophitY, sseqid == "Y2")) # 220 on Y2
# what is the proportion of query gene aligned to hap2?
#tophitY_prop<- tophitY #%>% mutate(alignedpropn = length/qlen) 
## filter out genes that have less than half of their sequence length aligned to hap2
tophitY_partialloss <- tophitY %>% filter(alignedpropn < 0.5)
tophitY_notloss <- tophitY %>% filter(alignedpropn > 0.5)

#print(tophitY_partialloss)
dim(tophitY_partialloss) # 371 genes
dim(filter(tophitY_partialloss, sseqid == "Y1")) # 168 on Y1
dim(filter(tophitY_partialloss, sseqid == "Y2")) # 203 on Y2
```
e.g. NChap1_final_00000186 has 3 hits to hap2 - one with 100% id, one with 98, one with 83. all on X

## control/opposite - what about hap1 gene loss
control-ish (aka is there X loss?)
```{r}
hap2_nonsyn<-data.table::fread("pg_hap2_wide.txt")

nsorths2<-hap2_nonsyn %>% filter_all(any_vars(grepl("\\*",.))) 

onlysyn2<-hap2_raw %>% filter(!(pgID%in%nsorths$pgID)) # 30505 genes remain?


onlysyn_sal<-sal_raw %>% filter(!(pgID%in%nsorths$pgID)) # 29906?
hasanortho<-onlysyn %>% filter((hap1!=""|hap2!="")&salicifolius!="")

hap1_yloss_redo<-hasanortho %>% filter(hap1!="" & hap2=="" & salicifolius!="")  #number doesn't change (expected)

hap2_xloss<-hasanortho %>% filter(hap1=="" & hap2!="" & salicifolius!="")  #number doesn't change (expected)

a_xloss<-hap2_xloss %>% group_by(interpChr) %>% summarise(n=n()) %>% 
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

b_xloss<-hasanortho %>% group_by(interpChr) %>% summarise(n=n()) %>%
  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" |interpChr=="X")

pars_xloss <- hap2_xloss %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_xloss_pars <-pars_xloss %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

total_pangenes_redo<-hasanortho %>% mutate(region=if_else(interpChr =="X" & end<=71000000,"PAR1",
                        if_else(interpChr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(interpChr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(interpChr=="X" & end>367000000,"PAR2",
                        if_else(interpChr=="A1","A1",
                        if_else(interpChr=="A2","A2",
                        if_else(interpChr=="A4","A4",NA))))))))
summary_total_pangenes_redo <-total_pangenes_redo %>% group_by(region) %>% summarise(n=n()) %>% 
  filter(!is.na(region))

chrpropnPAR_xloss<-full_join(summary_total_pangenes_redo,summary_xloss_pars,by="region",suffix=c(".total",".notonhap1")) %>% 
   mutate(., propn=n.notonhap1/n.total) %>% mutate(.,percent = propn*100) %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 
## plot proportions 

ggplot(chrpropnPAR_redo) + geom_col(aes(x=region,y=propn)) + 
  theme_classic() + 
  ylab("proportion annotated pangenes\nshared by haplotype 1 and salicifolius\n with no haplotype 2 ortholog")
ggsave("pangenepropn_yloss_redo.png", width=6,height=4) 
```



