---
title: "Presence/absence variation among Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

# data cleaning
Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(vroom)
```

## raw files
Load in long-format pangenome databases for each reference
```{r echo=TRUE, message=FALSE}
hap1raw<-vroom("hap1_pangenes.txt", show_col_types = F)
hap2raw<-vroom("hap2_pangenes.txt",show_col_types=F)
salraw<-vroom("salicifolius_pangenes.txt", show_col_types=F)
```
What do these column names mean?

colname  |definition
---------|------
ofID | orthofinder-given unique gene ID
pgID | the unique position-by-orthogroup combination identifier. Each row in the pangenome-annotation is a unique value of this field
interpChr |  the chromosome (syntenic or actual) on the reference genome
interpOrd     | the gene-rank order position (interpolated or actual) on the reference genome
pgRepID     | the ofID for the representative gene for each pgID
genome    | genome of the focal gene (id column)
og   | orthogroup ID taken from the combBed.txt file
flag| specifies whether the gene is a syntenic ortholog (‘PASS’), non-syntenic ortholog, or array member.
id| bed-specified gene ID 
chr| chromosome
start| start pos
end | end pos
ord | gene rank order

### what do I need to do?
```{r}
hap1syn<- hap1raw %>% dplyr::filter(flag=="PASS")
hap2syn<-hap2raw %>% dplyr::filter(flag=="PASS")
salsyn<-salraw %>% dplyr::filter(flag=="PASS")
```
Determine which genes are present in haplotype 1 and present or absent in haplotype 2
Determine which genes are present in haplotype 2 and not presetn in absent in haplotype 1
Determine which genes are not shared by sal and the other two, or are only shared between sal and one or the other. 
### what will the data look like
Here I'm envisioning a table of shared positions and gene names (column separated by species. I can include a flag indication which don't have a syntenic ortholog. 
simplest candidate for Y gene loss: single copy syntenic hap1 and sal but not in hap2.
What is my ultimate question? 
Is there gene loss on Y? neo Y? 
### Gene loss on Y

#### Syntenic genes
Obtain a list of genes shared by hap1, sal, but NOT hap2


##### bad attempt?
```{r}
## to keep and combine
library(tidyr)
spec1<- tibble(.name=c("hap1","hap2","salicifolius"),
               .value=c(""))
wide_data_hap1 <- pivot_wider(hap1syn, names_from = genome, values_from = pgID, 
                         names_prefix = "is", values_fill = list(pgID = FALSE))

rm(pg)

for i in pgID

(tidyr)




bool_hap1<- hap1syn %>% mutate(isHap1=ifelse(genome%in%"hap1",TRUE, FALSE)) %>%
  mutate(isHap2=ifelse(genome%in%"hap2",TRUE,FALSE)) %>%  mutate(isSal=ifelse(genome%in%"salicifolius",TRUE,FALSE))

wide_df <- bool_hap1 %>% pivot_wider(names_from = genome, values_from = id)

MissingInHap2<-bool_hap1 %>% group_by(pgID) %>% filter(isHap1 == TRUE&isHap2==FALSE)
SharedInHap2<- bool_hap1 %>% group_by(pgID) %>% filter(isHap1 == TRUE&isHap2==TRUE)


groupedtest<- hap1syn %>% group_by(pgID) %>% summarise(hasHap1=ifelse(genome%in%"hap1",TRUE, FALSE)) %>%   summarise(hap1syn,hasHap2=ifelse(genome%in%"hap2",TRUE, FALSE))
```
