---
title: "Single-copy genes missing and present from Rumex xyy male haplotype assemblies and Rumex salicifolius"
output:
  rmdformats::html_clean:
    code_folding: show
    default_style: light
    toc_depth: 6
    fig_caption: yes
---

### Setup and data cleaning

Goal: obtain a list of 1-1 syntenic orthologs that are present in hap1 and sal, but not hap2 - as well as the opposite case. Particularly interested in the sex chromosomes.

```{r message=FALSE, warning=FALSE}
library(renv)
library(tidyverse)
library(vroom)
library(GENESPACE)
library(googlesheets4)
setwd("~/Dropbox/rumex_pangenome_annotation")
```
#### raw files
Loading in wide format from query_pangenes() for each possible reference and bed files.
```{r echo=TRUE, message=FALSE}
# excluded non representative array members and non-syntenic othogroups (see genespace script)
hap1_raw<-data.table::fread("pg_hap1_wide_synonly.txt")
hap2_raw<-data.table::fread("pg_hap2_wide_synonly.txt")
hap1_nonsyn<-data.table::fread("pg_hap1_wide.txt")
```

### Filtering pangenome IDs w NS orths
Eliminating pangenome ID groups w non-syn orthogroup members. This is to get a more conservative set of "gene loss" candidates. NS orths are still of interest will be analyzed separately.
```{r}
# NSorths in the hap1_nonsyn file are annotated with a *
# creating a list with grepl - any entry with a *
nsorths<-hap1_nonsyn %>% filter_all(any_vars(grepl("\\*",.))) 
# filtering out any pgIDs in that set
onlysyn<-hap1_raw %>% filter(!(pgID%in%nsorths$pgID)) # 31289 genes remain
# same with salicifolius
#onlysyn_sal<-sal_raw %>% filter(!(pgID%in%nsorths$pgID)) # 29906
```

### Gene loss on Y
Identify genes that are shared by salicifolius and hap1, but not hap2. Adding position of PARS and correct denominator for calculating proportions.
```{r}
hasanortho<-onlysyn %>% 
  filter((hap1!=""|hap2!="")&salicifolius!="")
#write_delim(hasanortho,"denomgenelist.txt",delim="\t")
#write_delim(hasanortho_justsal,"denomgenelist_salorth.txt",delim="\t")
#write_csv(salorths, "denomgenelist_salorths.csv")
hap1_yloss<-onlysyn %>% filter(hap1!="" & hap2=="" & salicifolius!="")  

#######################################
###--- functions----
#######################################
## change slr region label
# major change from prev version: use chr instead of interpChr (for more accurate placement of SLRs)
slr_rename_h1<-function(x){mutate(x,region=if_else(chr =="X" & end<=71000000,"PAR1",
                        if_else(chr =="X" & (end>71000000 & end<=261000000),"OldX",
                        if_else(chr=="X" & (end>261000000 & end<=367000000),"NewX", 
                        if_else(chr=="X" & end>367000000,"PAR2",
                        if_else(chr=="A1","A1",
                        if_else(chr=="A2","A2",
                        if_else(chr=="A4","A4",NA))))))))}
yloss_pars<-slr_rename_h1(hap1_yloss)
# summarise numbers by region
summarisebyregion<-function(x,y){ x %>% tibble() %>%
  group_by(region) %>% summarise({{y}}:=n()) %>% 
  filter(!is.na(region))}
summary_yloss_pars<-summarisebyregion(yloss_pars,n_yloss)

total_pangenes<-slr_rename_h1(hasanortho)
summary_total_pangenes<-summarisebyregion(total_pangenes,n_total)
```

#### results so far

genes missing from hap2 
#### list for blasting genes against hap2 to confirm their loss

```{r}
genelisth1<-hap1_yloss %>% dplyr::select(hap1)
genelisth1_chr<-hap1_yloss %>% dplyr::select(hap1,interpChr)
#genelistsal<-hap1_yloss_redo %>% select(salicifolius)
#write.table(genelisth1,"genelisthap1.txt",quote=FALSE,sep="\t",row.names=F)
#write.table(genelistsal, "genelistsal.txt",quote=FALSE,sep="\t")
```

##### BLAST

Run in ohta because stuff is installed there

```{bash eval=TRUE, error=FALSE, warning=FALSE}
# ./blast_genome.sh
cat blast_genome.sh
```

#### futher analysis of blast output in R

```{r, warning=FALSE,message=FALSE}
blastout<-read_tsv("blast_h2genome.tsv",
                    col_names = c("qseqid", "sseqid", "pident", "length", 
                                  "mismatch", "gapopen", "qstart", "qend", 
                                  "qlen", "sstart", "send", "slen", "evalue", 
                                  "bitscore", "score"))

blast_pos<-full_join(blastout,hap2_raw,by=c("qseqid"="hap1"))  %>% add_count(qseqid) 
### which genes, out of any, are no hits in hap2 genome
hits<-blastout %>% dplyr::select(qseqid) %>% distinct()
nohit<- anti_join(genelisth1,hits, by = c("hap1"="qseqid"))
tophit_geneonly<- blastout %>% group_by(qseqid) %>%
  mutate(the_rank  = rank(-pident, ties.method = "random")) %>%
  filter(the_rank == 1) %>% dplyr::select(-the_rank) %>% mutate(alignedpropn = length/qlen)          
```

### exclude paralogs from blast top hit list

```{r}
tophit_samechr<- tophit_geneonly %>%
  left_join(genelisth1_chr,by=c("qseqid"="hap1")) %>%  
  mutate(matchchr=if_else(interpChr == "X", "Y",
                           if_else(interpChr == "A1","A1",
                           if_else(interpChr == "A2","A2",
                           if_else(interpChr == "A4","A4",NA))))) %>%
  filter(str_detect(sseqid,matchchr))
tophit_partial<- tophit_samechr %>% filter(alignedpropn <= 0.5)
tophit_notloss<-tophit_samechr %>% filter(alignedpropn > 0.5)

updated_genelist <- genelisth1 %>% filter(!hap1%in%tophit_notloss$qseqid)  
#write.table(updated_genelist,"genelisthap1_blast.txt",quote=FALSE,sep="\t",row.names=F)

```

```{r}
yloss_blast<- slr_rename_h1((hap1_yloss %>% filter(!hap1%in%tophit_notloss$qseqid)))
summary_yloss_blast <-summarisebyregion(yloss_blast,n_blastconfirmedloss)

yloss_partial <- slr_rename_h1((filter(hap1_yloss,hap1%in%tophit_partial$qseqid))) 
summary_yloss_partial <-summarisebyregion(yloss_partial,n_partialloss)
#write.csv(yloss_partial,"yloss_partial_list.csv")
## create complete loss file
complete_yloss<- slr_rename_h1((hap1_yloss %>% filter(!hap1%in%tophit_notloss$qseqid))) %>% filter(!hap1%in%tophit_partial$qseqid)
#write.csv(complete_yloss,"complete_yloss.csv")
#sheet<-gs4_create("yloss_partial_list",sheets =yloss_partial)
retained_on_y<-slr_rename_h1((hap1_yloss %>% filter(hap1%in%tophit_notloss$qseqid)))
#write.csv(retained_on_y,"retained_y.csv")


mylist<-list(summary_total_pangenes,summary_yloss_pars,
             summary_yloss_blast,summary_yloss_partial)

combinedsummary<-mylist %>% purrr::reduce(left_join,by="region")
propnsummary<-combinedsummary %>%
  mutate(propn_yloss=n_yloss/n_total, 
         propn_blastconfirmedloss=n_blastconfirmedloss/n_total,
         propn_partiallosstot=n_partialloss/n_total, 
         propn_partiallossbl=n_partialloss/n_blastconfirmedloss) %>% 
  mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldX","NewX","PAR2"))) 

print(propnsummary)
#sheet1 <- gs4_create("yloss_summary_jun23_23", sheets = propnsummary)

#write_delim(propnsummary,"summary_yloss.txt")
```

e.g. NChap1_final_00000186 has 3 hits to hap2 - one with 100% id, one with 98, one with 83. all on X

## control/opposite - what about hap1 gene loss

control-ish (aka is there X loss?)

```{r}
hap2_nonsyn<-data.table::fread("pg_hap2_wide.txt")
nsorths2<-hap2_nonsyn %>% filter_all(any_vars(grepl("\\*",.)))

onlysyn2<-hap2_nonsyn %>% filter(!(pgID%in%nsorths2$pgID)) # 30505 genes remain?
#
hasanortho2<-onlysyn2 %>%
   filter(((hap1!=""|hap2!="")&salicifolius!=""))
hap2_xloss<-hasanortho2 %>% filter(hap1=="" & hap2!="" & salicifolius!="")

#a_xloss<-hap2_xloss %>% group_by(interpChr) %>% summarise(n=n()) %>%
#  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" #|interpChr=="Y1"|interpChr=="Y2")

#b_xloss<-hasanortho %>% group_by(interpChr) %>% summarise(n=n()) %>%
#  filter(interpChr== "A1"|interpChr=="A2"| interpChr=="A4" #|interpChr=="Y1"|interpChr=="Y2")

 slr_rename_h2<- function(x){mutate(x,
   region=if_else(chr=="Y1" & end<=64000000,"NewY1",
          if_else(chr =="Y1" & (end>64000000 & end<=272000000),"OldY1",
          if_else(chr =="Y1" & (end > 272000000 & end <= 344000000),"PAR1",
     if_else(chr=="Y2" & end<=121000000,"PAR2",
     if_else(chr=="Y2"& (end>121000000 & end<=159000000),"NewY2",
     if_else(chr=="Y2" & end>159000000, "OldY2",
     if_else(chr=="A1","A1",
     if_else(chr=="A2","A2",
     if_else(chr=="A4","A4",NA))))))))))}

pars_xloss<-slr_rename_h2(hap2_xloss)
 
summary_xloss_pars<-summarisebyregion(pars_xloss,n_xloss)

total_pangenes2<-slr_rename_h2(hasanortho2)
 
summary_total_pangenes2<-summarisebyregion(total_pangenes2,n_total)
 
# chrpropnPAR_xloss<-full_join(summary_total_pangenes2,summary_xloss_pars,by="region") %>% 
#    mutate(., propn=n_xloss/n_total)  %>%
#    mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldY1","NewY2","OldY2","NewY1","PAR2"))) 
# ## plot proportions 
# print(chrpropnPAR_xloss)
# #sheet2 <- gs4_create("xloss_summary_jun23_23", sheets = chrpropnPAR_xloss)
#write_delim(chrpropnPAR_xloss,"summary_xloss.txt")
```

```{r}
## redo b/c totals look really weird!
slr_rename_h2<- function(x){mutate(x,
  region=if_else(chr=="Y1" & end<=64000000,"NewY1",
         if_else(chr =="Y1" & (end>64000000 & end<=272000000),"OldY1",
         if_else(chr =="Y1" & (end > 272000000 & end <= 344000000),"PAR1",
    if_else(chr=="Y2" & end<=121000000,"PAR2",
    if_else(chr=="Y2"& (end>121000000 & end<=159000000),"NewY2",
    if_else(chr=="Y2" & end>159000000, "OldY2",
    if_else(chr=="A1","A1",
    if_else(chr=="A2","A2",
    if_else(chr=="A4","A4",NA))))))))))}

pars_xloss<-slr_rename_h2(hap2_xloss)

summary_xloss_pars_fix<-summarisebyregion(pars_xloss,n_xloss)

total_pangenes2<-slr_rename_h2(hasanortho2)

summary_total_pangenes2<-summarisebyregion(total_pangenes2,n_total)

chrpropnPAR_xloss<-full_join(summary_total_pangenes2,summary_xloss_pars,by="region") %>%
   mutate(., propn=n_xloss/n_total)  %>%
   mutate(region = factor(region, levels=c("A1","A2","A4","PAR1","OldY1","NewY2","OldY2","NewY1","PAR2"))) 
## plot proportions 
print(chrpropnPAR_xloss)


pars_yloss<-slr_rename_h1(hap1_yloss)
summary_yloss_pars<-summarisebyregion(pars_yloss,n_yloss)


```

### plotting y loss

pie chart? bar plot? lets see...

```{r}
df<-propnsummary %>% mutate(propn_notloss = 1)%>%
   select(1,7,8,10) %>% 
  pivot_longer(cols=2:4,values_to = "summary")

ggplot(df, aes(region,summary,fill = name)) + geom_col(position="fill")
### ick
### --- plot of numbers genes not lost, partially lost, and completely lost --- ###
df1<- propnsummary %>% 
  mutate(n_notloss = n_total - n_blastconfirmedloss) %>% 
  mutate(n_totallylost = n_blastconfirmedloss - n_partialloss) %>%
  select(1,11,5,10) %>%
  pivot_longer(cols=2:4, values_to = "number") 
ggplot(df1, aes(region, number, fill=name)) + geom_col(position="stack") +
  theme_classic() 
#ggsave("barplot_geneloss.jpg",dpi = 300, width = 6, height = 4, units = "in")
##"
#install.packages("wesanderson")
library(wesanderson)

library(cowplot)
### --- pie chart? --- ####
df1oldX<- filter(df1, region == "OldX")
df1newX<-filter(df1, region == "NewX")
df1auto<-filter(df1, str_detect(region,"A.")) %>% group_by(name) %>% summarise(number=sum(number))
dfPAR1<-filter(df1, region == "PAR1")
dfPAR2<-filter(df1, region == "PAR2")

a<-ggplot(df1auto, aes(x="", y=number, fill=name)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) + theme_void()
b<-ggplot(df1oldX, aes(x="", y=number, fill=name)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) + theme_void()
c<-ggplot(df1newX, aes(x="", y=number, fill=name)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) + theme_void()
d<-ggplot(dfPAR1, aes(x="", y=number, fill=name)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) + theme_void()
e<-ggplot(dfPAR2, aes(x="", y=number, fill=name)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) + theme_void()
```

Combine legend <https://wilkelab.org/cowplot/articles/shared_legends.html>

```{r}
prow <- plot_grid(
  a + theme(legend.position="none"),
  b + theme(legend.position="none"),
  c + theme(legend.position="none"),
  d + theme(legend.position = "none"),
  e + theme(legend.position = "none"),
  align = 'vh',
  labels = c("Autosomes","Old Y","New Y","PAR1","PAR2"),
  hjust = 0,
  label_y=0.3,
  nrow = 1
)


prow
legend <- cowplot::get_legend(a + 
              theme(legend.box.margin = margin(0, 0, 0, 25))+
              scale_fill_discrete(name="Paternal haplotype", 
              labels=c("Retained genes", "Partially lost genes", "Completely lost genes")))

# extract a legend that is laid out horizontally
plot_grid(prow, legend, rel_widths = c(3, 1))
ggsave("pie_geneloss_june23.jpg",dpi = 300, width = 10, height = 4, units = "in")


```

```{r}
testtest<-dplyr::setdiff(total_pangenes$salicifolius,total_pangenes2$salicifolius)
## 551 diff 
testhap1<-dplyr::setdiff(total_pangenes$hap1,total_pangenes2$hap1)
## 481
testhap1<-dplyr::setdiff(total_pangenes$hap1,total_pangenes2$hap1)

```

#### 

Pie plots are terrible, using suggestion from Zoe for bar plots

```{r}

```
